darwin: stress_mt hangs macOS CI (~20% of runs) — event-thread startup/shutdown handshake can wait forever

> **🤖 Analysis by Claude** — this investigation and write-up were generated by Claude (model `claude-fable-5`, via Claude Code) at a maintainer's request. Findings are from code analysis and CI-run forensics; no macOS hardware was available to reproduce locally.

> [!IMPORTANT]
> **Update (2026-07-12):** the "unlatched startup/shutdown handshake" hypothesis below, and the "CI runners are device-less" assumption it partly rests on, were **both wrong** — see the [follow-up comment](https://github.com/libusb/libusb/issues/1869#issuecomment-4950063680) for a real captured stack trace from a reproduced CI hang. The event thread turns out to be healthy; all 8 worker threads are wedged waiting on a lost transfer completion, matching @mcuee's TSan-confirmed `open_count`/`capture_count` race (#1798 item 1, fixed by #1864). Kept below for the record of how the investigation got there.

## Symptom

`stress_mt` hangs forever on the macOS GitHub CI, the job dies at the 6-hour GitHub timeout as `cancelled`, and all evidence is lost (automake buffers the test's output into `test-suite.log`, which is never shown).

Sampled forensics from [run 28896022560](https://github.com/libusb/libusb/actions/runs/28896022560) (branch `conversion-warnings`, 2026-07-08):

```
00:19:00  Making check in tests
00:19:12  PASS: stress
          (silence for 5h58m)
06:17:52  ##[error]The operation was canceled.
06:17:54  Terminate orphan process: pid (13341) (stress_mt)
```

**Frequency: of the last 100 macOS workflow runs, 21 were cancelled** — every one I sampled ran exactly ~6h before being killed, i.e. they are all this hang, including runs on `master` itself ([28894513259](https://github.com/libusb/libusb/actions/runs/28894513259), [28881034605](https://github.com/libusb/libusb/actions/runs/28881034605), [28819210881](https://github.com/libusb/libusb/actions/runs/28819210881)). Hangs were reported as far back as May 2024 in #1494, and the rate is essentially unchanged before/after #1780's merge (2026-07-02): cancelled runs appear continuously through April–July 2026 on both sides of that date.

## Constraint analysis — where the hang can and cannot be

A key structural fact narrows this a lot: **core serializes all backend init/exit.** `libusb_init_context` holds `default_context_lock` across `usbi_backend.init` ([core.c#L2597-L2711](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/core.c#L2597-L2711)) and `libusb_exit` holds it across `usbi_backend.exit` ([core.c#L2750-L2789](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/core.c#L2750-L2789)). Consequences:

- `darwin_init_context` and `darwin_exit` can never run concurrently with each other, so "overlapping event-thread epochs" (a second event thread created while the previous one is being torn down) are **impossible** — ruled out.
- In stress_mt phase 1 (pure init/exit) the whole test is effectively sequential except for one actor: **the darwin event thread**. Any hang must be in the interaction between exactly one user thread and that event thread.
- Once one thread hangs inside init/exit, it holds `default_context_lock` and **every other thread in the process freezes** at its next `libusb_init_context`/`libusb_exit` — matching the observed total silence with no partial progress and no test-failure output.

Also ruled out:

- #1780's deadlock (event-thread hotplug callback taking `darwin_cached_devices_mutex` during exit teardown) **cannot fire on the CI runners**: they expose no USB devices, so `darwin_devices_attached`/`darwin_devices_detached` never run. Consistent with the unchanged hang rate after #1780 was merged.
- `darwin_hotplug_poll` is a bounded 1 ms `IOKitWaitQuiet` — not a hang candidate.
- `darwin_default_master_port` is the constant default port — no per-epoch churn there.
- The pthread cond/mutex primitives and the signal-before-wait patterns are individually correct.

## The libusb-side defect: an unlatched, failure-blind startup handshake

Per init/exit cycle (stress_mt does **8 threads × 64 iterations × 2 phases ≈ 1024 epochs**), `darwin_first_time_init` creates the event thread and waits ([darwin_usb.c#L901-L918](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/os/darwin_usb.c#L901-L918)):

```c
  int rc = pthread_create (&libusb_darwin_at, NULL, darwin_event_thread_main, NULL);
  ...
  usbi_mutex_lock (&libusb_darwin_at_mutex);
  libusb_darwin_at_started = true;
  while (NULL == libusb_darwin_acfl) {
    usbi_cond_wait (&libusb_darwin_at_cond, &libusb_darwin_at_mutex);
  }
```

The event thread publishes and later clears **the same variable** ([darwin_usb.c#L830-L847](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/os/darwin_usb.c#L830-L847)):

```c
  /* publish */
  libusb_darwin_acfl = runloop;  ... usbi_cond_signal(...);
  CFRunLoopRun();
  /* clear */
  libusb_darwin_acfls = NULL;
  libusb_darwin_acfl = NULL;     ... usbi_cond_signal(...);
```

`libusb_darwin_acfl` encodes two different predicates — "startup finished" (for init, non-NULL) and "runloop still running" (for exit, NULL) — with no latch and no failure state other than the explicit `darwin_fail_startup()` paths. That leaves two hang shapes, both of which wedge the entire process forever while holding `default_context_lock`:

1. **Publish→clear flicker.** If `CFRunLoopRun()` returns prematurely (i.e. not via `darwin_exit`'s shutdown source), the event thread can execute *publish* and *clear* back-to-back. The waiter in `first_time_init` may only re-acquire the mutex after both transitions: it then observes `acfl == NULL`, re-enters `cond_wait`, and **no signal will ever come again** — permanent hang. Nothing in the handshake distinguishes "not yet published" from "published and already gone".
2. **Pre-publish wedge.** If the event thread blocks indefinitely inside its bring-up sequence — `IONotificationPortCreate`, 2× `IOServiceAddMatchingNotification`, iterator arming ([darwin_usb.c#L790-L828](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/os/darwin_usb.c#L790-L828)) — the waiter waits forever. There is no timeout and no error path.

The same unlatched pattern exists on the exit side (`while (libusb_darwin_acfl) cond_wait`, [darwin_usb.c#L963-L969](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/os/darwin_usb.c#L963-L969)), followed by an unconditional `pthread_join`.

## Likely trigger (needs one stack trace to confirm)

What makes `CFRunLoopRun()` return early or the IOKit bring-up wedge roughly once per several hundred epochs? Candidates, in order of suspicion:

- **CoreFoundation per-thread runloop lifecycle under rapid thread churn.** `CFRunLoopGetCurrent()` caches runloops keyed by `pthread_t`; entries are torn down via TLS destructors at thread exit. stress_mt recycles ~1024 short-lived event threads as fast as the machine allows, guaranteeing heavy `pthread_t` reuse — a classic way to hit stale/finalizing runloop objects (the code does `CFRunLoopGetCurrent` + `CFRetain` on every epoch, [darwin_usb.c#L783](https://github.com/libusb/libusb/blob/d20b5b41aa/libusb/os/darwin_usb.c#L783)).
- **IOKit notification-port churn**: `IONotificationPortCreate`/`IOServiceAddMatchingNotification`/`IONotificationPortDestroy` per epoch — Mach RPCs into IOKit at high frequency on a virtualized runner.
- Note the CI builds run with **ASan enabled** (`.private/ci-build.sh` defaults `asan=yes`), which perturbs timing/allocation and may widen the race window.

I could not pinpoint the exact CF/IOKit-internal trigger without a stack from a hung run — see "next step" below.

## Proposed fix direction (backend-local, modest size)

Replace the two-meanings-in-one-pointer handshake with an explicit latched state, e.g.:

```c
enum { DARWIN_AT_STOPPED, DARWIN_AT_STARTING, DARWIN_AT_RUNNING, DARWIN_AT_FAILED, DARWIN_AT_EXITED }
  libusb_darwin_at_state GUARDED_BY(libusb_darwin_at_mutex);
```

- Event thread: `STARTING → RUNNING` at publish, `→ FAILED` in `darwin_fail_startup`, `→ EXITED` after `CFRunLoopRun` returns; always `usbi_cond_broadcast`.
- `darwin_first_time_init`: wait `while (state == STARTING)`; treat anything but `RUNNING` as `LIBUSB_ERROR_OTHER` (join the thread before returning the error so nothing leaks).
- `darwin_exit`: signal the shutdown source only in `RUNNING`; wait `while (state == RUNNING)`; join as today.

This turns hang shape 1 into a clean `libusb_init` failure (visible red test instead of a 6-hour zombie job) and makes every state transition observable. Shape 2 (a genuine IOKit wedge) can additionally be bounded with `usbi_cond_timedwait` on the startup wait, returning `LIBUSB_ERROR_OTHER` after a generous timeout — debatable as library policy, but strictly better than wedging every libusb user in the process forever; opinions welcome.

I'm happy to turn this into a PR if the direction looks right to maintainers.

## Next step to confirm the trigger

Since the hang destroys its own evidence on CI, I'm proposing a small CI watchdog (separate PR, linked below) that, when `stress_mt` is still alive after 40 minutes, captures `sample`-based stacks of the hung process, dumps `test-suite.log`, and fails the job in under an hour instead of six. The first captured stack will confirm which of the two shapes (and which CF/IOKit call) we're dealing with.

## Related

- #1494 (stress_mt failures and stuck on macOS — the hangs reported there since 2024 match this analysis)
- #1798 (Darwin backend concurrency audit; items 4/8 are adjacent but do not explain a CI hang with no devices present)
- #1780 (fixed a real exit-path deadlock, but one that cannot occur on device-less CI runners — hang rate unchanged since)

Now I can 100% reproduce the run with libusb git under my MacBook Neo.

```
mcuee in ~/build/libusb/libusb on master % sw_vers 
ProductName:		macOS
ProductVersion:		26.5.2
BuildVersion:		25F84

mcuee in ~/build/libusb/libusb on master % cc -v
Apple clang version 21.0.0 (clang-2100.1.1.101)
Target: arm64-apple-darwin25.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

mcuee in ~/build/libusb/libusb on master % lsusb_macos 
Bus 000 Device 001: ID 05ac:8015 Apple USB2.0 Hub  Serial: Not Provided
Bus 000 Device 003: ID 05e3:0610 GenesysLogic USB2.1 Hub  Serial: Not Provided
Bus 000 Device 004: ID 03eb:2104 ATMEL AVRISP mkII  Serial: 001D2C990079
Bus 000 Device 007: ID 04b4:8613  Unnamed Device  Serial: Not Provided
Bus 000 Device 002: ID 05e3:0625 GenesysLogic USB3.2 Hub  Serial: Not Provided
Bus 000 Device 005: ID 0bda:8156 Realtek USB 10/100/1G/2.5G LAN  Serial: 4013000001
Bus 000 Device 006: ID 05e3:0747 Generic USB Storage  Serial: 000000000819
Bus 000 Device 001: ID 05ac:8007 Apple Inc. XHCI Root Hub USB 2.0 Simulation 

mcuee in ~/build/libusb/libusb on master % git pull
Already up to date.

mcuee in ~/build/libusb/libusb on master % git log --oneline -1
d20b5b41 (HEAD -> master, origin/master, origin/HEAD) Corrected undefined signed overflow

mcuee in ~/build/libusb/libusb on master % CFLAGS="-fsanitize=thread -g" ./autogen.sh --prefix=/Users/mcuee/mybin/bingit_tsan_debug
...
mcuee in ~/build/libusb/libusb on master % make
...
mcuee in ~/build/libusb/libusb on master % make install
...
mcuee in ~/build/libusb/libusb on master % make check
Making check in libusb
  GEN      regen-version-describe
/Library/Developer/CommandLineTools/usr/bin/make  check-am
  GEN      regen-version-describe
  CC       core.lo
  CCLD     libusb-1.0.la
Making check in examples
  CCLD     dpfp
  CCLD     dpfp_threaded
  CCLD     fxload
  CCLD     hotplugtest
  CCLD     listdevs
  CCLD     sam3u_benchmark
  CCLD     testlibusb
  CCLD     xusb
Making check in tests
  CCLD     stress
  CCLD     stress_mt
  CCLD     set_option
  CCLD     init_context
  CCLD     macos
/Library/Developer/CommandLineTools/usr/bin/make  check-TESTS
PASS: stress
^C (note: hang here)
make[3]: *** Deleting file `stress_mt.log'
make[3]: *** [stress_mt.log] Error 130
make[2]: *** [check-TESTS] Interrupt: 2
make[1]: *** [check-am] Interrupt: 2
make: *** [check-recursive] Interrupt: 2

mcuee in ~/build/libusb/libusb on master % ./tests/stress_mt 
Running multithreaded init/exit test...
Starting 8 threads
Running multithreaded init/exit test with enumeration...
Starting 8 threads
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 8 at 0x00010ac07f00 by thread T523:
    #0 usbi_get_transfer_priv libusbi.h:921 (stress_mt:arm64+0x1000243fc)
    #1 darwin_async_io_callback darwin_usb.c:2687 (stress_mt:arm64+0x10002447c)
    #2 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 8 at 0x00010ac07f00 by thread T522:
    #0 libusb_alloc_transfer io.c:1308 (stress_mt:arm64+0x100010398)
    #1 libusb_control_transfer sync.c:117 (stress_mt:arm64+0x100014760)
    #2 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Location is heap block of size 264 at 0x00010ac07e40 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 libusb_alloc_transfer io.c:1303 (stress_mt:arm64+0x100010358)
    #2 libusb_control_transfer sync.c:117 (stress_mt:arm64+0x100014760)
    #3 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

SUMMARY: ThreadSanitizer: data race libusbi.h:921 in usbi_get_transfer_priv
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 4 at 0x00010a60835c by thread T524:
    #0 darwin_open darwin_usb.c:1601 (stress_mt:arm64+0x1000174f4)
    #1 libusb_open core.c:1490 (stress_mt:arm64+0x10000395c)
    #2 init_and_exit stress_mt.c:142 (stress_mt:arm64+0x100000e88)

  Previous write of size 4 at 0x00010a60835c by thread T522:
    #0 darwin_open darwin_usb.c:1643 (stress_mt:arm64+0x10001790c)
    #1 libusb_open core.c:1490 (stress_mt:arm64+0x10000395c)
    #2 init_and_exit stress_mt.c:142 (stress_mt:arm64+0x100000e88)

  Location is heap block of size 120 at 0x00010a608300 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 darwin_get_cached_device darwin_usb.c:1418 (stress_mt:arm64+0x10001c740)
    #2 darwin_scan_devices darwin_usb.c:1580 (stress_mt:arm64+0x10001a540)
    #3 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #4 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #5 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #6 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T524 (tid=122136, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000ae0)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:1601 in darwin_open
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Write of size 4 at 0x00010a60835c by thread T526:
    #0 darwin_open darwin_usb.c:1643 (stress_mt:arm64+0x10001790c)
    #1 libusb_open core.c:1490 (stress_mt:arm64+0x10000395c)
    #2 init_and_exit stress_mt.c:142 (stress_mt:arm64+0x100000e88)

  Previous write of size 4 at 0x00010a60835c by thread T525:
    #0 darwin_open darwin_usb.c:1643 (stress_mt:arm64+0x10001790c)
    #1 libusb_open core.c:1490 (stress_mt:arm64+0x10000395c)
    #2 init_and_exit stress_mt.c:142 (stress_mt:arm64+0x100000e88)

  Location is heap block of size 120 at 0x00010a608300 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 darwin_get_cached_device darwin_usb.c:1418 (stress_mt:arm64+0x10001c740)
    #2 darwin_scan_devices darwin_usb.c:1580 (stress_mt:arm64+0x10001a540)
    #3 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #4 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #5 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #6 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T526 (tid=122138, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b28)

  Thread T525 (tid=122137, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b04)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:1643 in darwin_open
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 8 at 0x00010ac07ef8 by thread T523:
    #0 darwin_async_io_callback darwin_usb.c:2689 (stress_mt:arm64+0x100024490)
    #1 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 8 at 0x00010ac07ef8 by thread T522:
    #0 libusb_submit_transfer io.c:1501 (stress_mt:arm64+0x1000106dc)
    #1 libusb_control_transfer sync.c:135 (stress_mt:arm64+0x100014838)
    #2 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Location is heap block of size 264 at 0x00010ac07e40 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 libusb_alloc_transfer io.c:1303 (stress_mt:arm64+0x100010358)
    #2 libusb_control_transfer sync.c:117 (stress_mt:arm64+0x100014760)
    #3 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:2689 in darwin_async_io_callback
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 8 at 0x00010a610c08 by thread T523:
    #0 darwin_async_io_callback darwin_usb.c:2689 (stress_mt:arm64+0x1000244c4)
    #1 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 8 at 0x00010a610c08 by thread T525 (mutexes: write M0):
    #0 usbi_alloc_device core.c:735 (stress_mt:arm64+0x100001640)
    #1 process_new_device darwin_usb.c:1496 (stress_mt:arm64+0x10001cc98)
    #2 darwin_scan_devices darwin_usb.c:1586 (stress_mt:arm64+0x10001a618)
    #3 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #4 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #5 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #6 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Location is heap block of size 120 at 0x00010a610c00 allocated by thread T525:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 usbi_alloc_device core.c:728 (stress_mt:arm64+0x1000015f8)
    #2 process_new_device darwin_usb.c:1496 (stress_mt:arm64+0x10001cc98)
    #3 darwin_scan_devices darwin_usb.c:1586 (stress_mt:arm64+0x10001a618)
    #4 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #5 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #6 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #7 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Mutex M0 (0x000100d00040) created at:
    #0 pthread_mutex_lock <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x356bc)
    #1 usbi_mutex_static_lock threads_posix.h:35 (stress_mt:arm64+0x10000608c)
    #2 libusb_init_context core.c:2597 (stress_mt:arm64+0x100006208)
    #3 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T525 (tid=122137, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b04)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:2689 in darwin_async_io_callback
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Write of size 4 at 0x00010a60835c by thread T522 (mutexes: write M0):
    #0 darwin_close darwin_usb.c:1662 (stress_mt:arm64+0x100017ab8)
    #1 do_close core.c:1609 (stress_mt:arm64+0x100004140)
    #2 libusb_close core.c:1663 (stress_mt:arm64+0x100003dc4)
    #3 init_and_exit stress_mt.c:200 (stress_mt:arm64+0x100000e3c)

  Previous write of size 4 at 0x00010a60835c by thread T524:
    #0 darwin_open darwin_usb.c:1643 (stress_mt:arm64+0x10001790c)
    #1 libusb_open core.c:1490 (stress_mt:arm64+0x10000395c)
    #2 init_and_exit stress_mt.c:142 (stress_mt:arm64+0x100000e88)

  Location is heap block of size 120 at 0x00010a608300 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 darwin_get_cached_device darwin_usb.c:1418 (stress_mt:arm64+0x10001c740)
    #2 darwin_scan_devices darwin_usb.c:1580 (stress_mt:arm64+0x10001a540)
    #3 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #4 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #5 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #6 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Mutex M0 (0x00010be0bb88) created at:
    #0 pthread_mutex_init <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x352ec)
    #1 usbi_mutex_init threads_posix.h:45 (stress_mt:arm64+0x10000fa7c)
    #2 usbi_io_init io.c:1167 (stress_mt:arm64+0x10000f8e4)
    #3 libusb_init_context core.c:2681 (stress_mt:arm64+0x1000067bc)
    #4 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

  Thread T524 (tid=122136, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000ae0)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:1662 in darwin_close
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 4 at 0x00010be03c00 by thread T523:
    #0 log_v core.c:2935 (stress_mt:arm64+0x100006fe8)
    #1 usbi_log core.c:3033 (stress_mt:arm64+0x1000014d0)
    #2 darwin_async_io_callback darwin_usb.c:2689 (stress_mt:arm64+0x1000244fc)
    #3 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 4 at 0x00010be03c00 by thread T525 (mutexes: write M0):
    #0 libusb_init_context core.c:2621 (stress_mt:arm64+0x10000634c)
    #1 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Location is heap block of size 752 at 0x00010be03c00 allocated by thread T525:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 libusb_init_context core.c:2614 (stress_mt:arm64+0x100006318)
    #2 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Mutex M0 (0x000100d00040) created at:
    #0 pthread_mutex_lock <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x356bc)
    #1 usbi_mutex_static_lock threads_posix.h:35 (stress_mt:arm64+0x10000608c)
    #2 libusb_init_context core.c:2597 (stress_mt:arm64+0x100006208)
    #3 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T525 (tid=122137, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b04)

SUMMARY: ThreadSanitizer: data race core.c:2935 in log_v
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Write of size 4 at 0x00010a60835c by thread T527 (mutexes: write M0):
    #0 darwin_close darwin_usb.c:1662 (stress_mt:arm64+0x100017ab8)
    #1 do_close core.c:1609 (stress_mt:arm64+0x100004140)
    #2 libusb_close core.c:1663 (stress_mt:arm64+0x100003dc4)
    #3 init_and_exit stress_mt.c:200 (stress_mt:arm64+0x100000e3c)

  Previous write of size 4 at 0x00010a60835c by thread T525 (mutexes: write M1):
    #0 darwin_close darwin_usb.c:1662 (stress_mt:arm64+0x100017ab8)
    #1 do_close core.c:1609 (stress_mt:arm64+0x100004140)
    #2 libusb_close core.c:1663 (stress_mt:arm64+0x100003dc4)
    #3 init_and_exit stress_mt.c:200 (stress_mt:arm64+0x100000e3c)

  Location is heap block of size 120 at 0x00010a608300 allocated by thread T522:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 darwin_get_cached_device darwin_usb.c:1418 (stress_mt:arm64+0x10001c740)
    #2 darwin_scan_devices darwin_usb.c:1580 (stress_mt:arm64+0x10001a540)
    #3 darwin_init_context darwin_usb.c:936 (stress_mt:arm64+0x10001a13c)
    #4 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #5 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #6 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Mutex M0 (0x00010be0fa88) created at:
    #0 pthread_mutex_init <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x352ec)
    #1 usbi_mutex_init threads_posix.h:45 (stress_mt:arm64+0x10000fa7c)
    #2 usbi_io_init io.c:1167 (stress_mt:arm64+0x10000f8e4)
    #3 libusb_init_context core.c:2681 (stress_mt:arm64+0x1000067bc)
    #4 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Mutex M1 (0x00010be03d88) created at:
    #0 pthread_mutex_init <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x352ec)
    #1 usbi_mutex_init threads_posix.h:45 (stress_mt:arm64+0x10000fa7c)
    #2 usbi_io_init io.c:1167 (stress_mt:arm64+0x10000f8e4)
    #3 libusb_init_context core.c:2681 (stress_mt:arm64+0x1000067bc)
    #4 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T527 (tid=122139, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b4c)

  Thread T525 (tid=122137, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b04)

  Thread T522 (tid=122134, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000abc)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:1662 in darwin_close
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 1 at 0x00010ac17e10 by thread T523:
    #0 darwin_async_io_callback darwin_usb.c:2692 (stress_mt:arm64+0x100024538)
    #1 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 1 at 0x00010ac17e10 by thread T524:
    #0 libusb_control_transfer sync.c:134 (stress_mt:arm64+0x100014824)
    #1 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Location is heap block of size 264 at 0x00010ac17d40 allocated by thread T524:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 libusb_alloc_transfer io.c:1303 (stress_mt:arm64+0x100010358)
    #2 libusb_control_transfer sync.c:117 (stress_mt:arm64+0x100014760)
    #3 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T524 (tid=122136, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000ae0)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:2692 in darwin_async_io_callback
==================
==================
WARNING: ThreadSanitizer: data race (pid=29056)
  Read of size 1 at 0x00010ac0bed1 by thread T523:
    #0 darwin_async_io_callback darwin_usb.c:2692 (stress_mt:arm64+0x100024518)
    #1 IODispatchCalloutFromCFMessage <null> (IOKit:arm64e+0x5334)

  Previous write of size 1 at 0x00010ac0bed1 by thread T525:
    #0 libusb_fill_control_transfer libusb.h:1996 (stress_mt:arm64+0x100014ca4)
    #1 libusb_control_transfer sync.c:132 (stress_mt:arm64+0x100014814)
    #2 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Location is heap block of size 264 at 0x00010ac0be00 allocated by thread T525:
    #0 calloc <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x67c40)
    #1 libusb_alloc_transfer io.c:1303 (stress_mt:arm64+0x100010358)
    #2 libusb_control_transfer sync.c:117 (stress_mt:arm64+0x100014760)
    #3 init_and_exit stress_mt.c:164 (stress_mt:arm64+0x100000ee0)

  Thread T523 (tid=122135, running) created by thread T522 at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 darwin_first_time_init darwin_usb.c:901 (stress_mt:arm64+0x10001a310)
    #2 darwin_init_context darwin_usb.c:928 (stress_mt:arm64+0x10001a0fc)
    #3 darwin_init darwin_usb.c:940 (stress_mt:arm64+0x100016ef0)
    #4 libusb_init_context core.c:2690 (stress_mt:arm64+0x100006844)
    #5 init_and_exit stress_mt.c:120 (stress_mt:arm64+0x100000de0)

  Thread T525 (tid=122137, running) created by main thread at:
    #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x33834)
    #1 test_multi_init stress_mt.c:224 (stress_mt:arm64+0x100000b04)

SUMMARY: ThreadSanitizer: data race darwin_usb.c:2692 in darwin_async_io_callback
==================
^C
```

--- a/libusb/os/darwin_usb.c
+++ b/libusb/os/darwin_usb.c
@@ -468,7 +468,14 @@ static void darwin_deref_cached_device(struct darwin_cached_device *cached_dev) 
       (*cached_dev->device)->Release(cached_dev->device);
       cached_dev->device = NULL;
     }
+    if (cached_dev->pending_device) {
+      (*cached_dev->pending_device)->Release(cached_dev->pending_device);
+      cached_dev->pending_device = NULL;
+    }
     IOObjectRelease (cached_dev->service);
+    if (cached_dev->pending_service != IO_OBJECT_NULL) {
+      IOObjectRelease (cached_dev->pending_service);
+    }
     usbi_mutex_destroy(&cached_dev->lock);
     free (cached_dev);
   }
@@ -476,6 +483,21 @@ static void darwin_deref_cached_device(struct darwin_cached_device *cached_dev) 
 
 static void darwin_ref_cached_device(struct darwin_cached_device *cached_dev) REQUIRES(darwin_cached_devices_mutex) {
   cached_dev->refcount++;
+}
+
+/* returns the live device interface for enumeration purposes: while the
+   device is re-enumerating, the replacement discovered by the hotplug thread
+   (pending_device) is the interface that works; the one in device is the old,
+   no longer functional one that the re-enumerating thread still owns. must
+   NOT be called with darwin_cached_devices_mutex held. */
+static usb_device_t darwin_active_device_interface (struct darwin_cached_device *cached_dev) EXCLUDES(darwin_cached_devices_mutex) {
+  usb_device_t device;
+
+  usbi_mutex_lock(&darwin_cached_devices_mutex);
+  device = cached_dev->pending_device ? cached_dev->pending_device : cached_dev->device;
+  usbi_mutex_unlock(&darwin_cached_devices_mutex);
+
+  return device;
 }
 
 static int ep_to_pipeRef(struct libusb_device_handle *dev_handle, uint8_t ep, uint8_t *pipep, uint8_t *ifcp, struct darwin_interface **interface_out) {
@@ -696,12 +718,6 @@ static void darwin_devices_detached (void *ptr, io_iterator_t rem_devices) EXCLU
            * will deref if needed. */
           usbi_dbg (NULL, "detected device detached due to re-enumeration. sessionID: 0x%" PRIx64
                           ", locationID: 0x%" PRIx32, session, locationID);
-
-          /* the device object is no longer usable so go ahead and release it */
-          if (old_device->device) {
-            (*old_device->device)->Release(old_device->device);
-            old_device->device = NULL;
-          }
 
           is_reenumerating = true;
         } else {
@@ -1107,7 +1123,7 @@ static int darwin_get_config_descriptor(struct libusb_device *dev, uint8_t confi
 
 /* check whether the os has configured the device */
 static enum libusb_error darwin_check_configuration (struct libusb_context *ctx, struct darwin_cached_device *dev) {
-  usb_device_t darwin_device = dev->device;
+  usb_device_t darwin_device = darwin_active_device_interface (dev);
 
   IOUSBConfigurationDescriptorPtr configDesc;
   IOUSBFindInterfaceRequest request;
@@ -1192,8 +1208,10 @@ static IOReturn darwin_request_descriptor (usb_device_t device, UInt8 desc, UInt
   return (*device)->DeviceRequestTO (device, &req);
 }
 
-static enum libusb_error darwin_cache_device_descriptor (struct libusb_context *ctx, struct darwin_cached_device *dev) {
-  usb_device_t device = dev->device;
+/* device is passed explicitly because this is called from
+   darwin_get_cached_device with darwin_cached_devices_mutex held, before the
+   pending replacement (if any) has been adopted into dev->device */
+static enum libusb_error darwin_cache_device_descriptor (struct libusb_context *ctx, struct darwin_cached_device *dev, usb_device_t device) {
   int retries = 1;
   long delay = 30000; /* microseconds */
   int unsuspended = 0, try_unsuspend = 1, try_reconfigure = 1;
@@ -1445,9 +1463,6 @@ static enum libusb_error darwin_get_cached_device(struct libusb_context *ctx, io
       new_device->parent_session = parent_sessionID;
 
       usbi_mutex_init(&new_device->lock);
-    } else {
-      /* release the ref to old device's service */
-      IOObjectRelease (new_device->service);
     }
 
     /* keep track of devices regardless of if we successfully enumerate them to
@@ -1455,14 +1470,29 @@ static enum libusb_error darwin_get_cached_device(struct libusb_context *ctx, io
     *cached_out = new_device;
 
     new_device->session = sessionID;
-    new_device->device = device;
-    new_device->service = service;
+    if (*old_session_id) {
+      /* the device is being re-enumerated: the re-enumerating thread may
+         still be using device and service, so do not touch them here. stash
+         the replacements instead; darwin_reenumerate_device adopts them (and
+         releases the old ones) once the re-enumeration completes. */
+      if (new_device->pending_device) {
+        (*new_device->pending_device)->Release (new_device->pending_device);
+      }
+      if (new_device->pending_service != IO_OBJECT_NULL) {
+        IOObjectRelease (new_device->pending_service);
+      }
+      new_device->pending_device = device;
+      new_device->pending_service = service;
+    } else {
+      new_device->device = device;
+      new_device->service = service;
+    }
 
     /* retain the service */
     IOObjectRetain (service);
 
     /* cache the device descriptor */
-    ret = darwin_cache_device_descriptor(ctx, new_device);
+    ret = darwin_cache_device_descriptor(ctx, new_device, device);
     if (ret)
       break;
 
@@ -1540,7 +1570,8 @@ static enum libusb_error process_new_device (struct libusb_context *ctx, struct 
       dev->parent_dev = usbi_get_device_by_session_id (ctx, (unsigned long) cached_device->parent_session);
     }
 
-    (*priv->dev->device)->GetDeviceSpeed (priv->dev->device, &devSpeed);
+    usb_device_t darwin_device = darwin_active_device_interface (priv->dev);
+    (*darwin_device)->GetDeviceSpeed (darwin_device, &devSpeed);
 
     switch (devSpeed) {
     case kUSBDeviceSpeedLow: dev->speed = LIBUSB_SPEED_LOW; break;
@@ -2201,6 +2232,33 @@ static enum libusb_error darwin_restore_state (struct libusb_device_handle *dev_
   return LIBUSB_SUCCESS;
 }
 
+/* adopt the replacement device interface and service discovered by the
+   hotplug thread during re-enumeration (see darwin_get_cached_device), and
+   release the old ones. after this, device and service have only ever been
+   written by threads holding dpriv->lock. must be called while holding
+   dpriv->lock. */
+static void darwin_adopt_pending_device (struct darwin_cached_device *dpriv) EXCLUDES(darwin_cached_devices_mutex) {
+  usbi_mutex_lock(&darwin_cached_devices_mutex);
+
+  if (dpriv->pending_device) {
+    if (dpriv->device) {
+      (*dpriv->device)->Release (dpriv->device);
+    }
+    dpriv->device = dpriv->pending_device;
+    dpriv->pending_device = NULL;
+  }
+
+  if (dpriv->pending_service != IO_OBJECT_NULL) {
+    if (dpriv->service != IO_OBJECT_NULL) {
+      IOObjectRelease (dpriv->service);
+    }
+    dpriv->service = dpriv->pending_service;
+    dpriv->pending_service = IO_OBJECT_NULL;
+  }
+
+  usbi_mutex_unlock(&darwin_cached_devices_mutex);
+}
+
 /* must be called while holding dpriv->lock. the lock is held during the wait
    for re-enumeration to complete, which serializes open/close/capture
    operations on the device while it is being reset. the hotplug thread that
@@ -2220,6 +2278,10 @@ static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, b
 
   struct libusb_context *ctx = HANDLE_CTX (dev_handle);
 
+  /* adopt any replacement left over from a previous re-enumeration that
+     returned (e.g. on timeout) before the device came back */
+  darwin_adopt_pending_device (dpriv);
+
   if (!dpriv->device) {
     usbi_warn(ctx, "darwin_reenumerate_device: device interface is NULL");
     return LIBUSB_ERROR_NO_DEVICE;
@@ -2230,8 +2292,13 @@ static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, b
     return LIBUSB_ERROR_NOT_FOUND;
   }
 
-  /* store copies of descriptors so they can be compared after the reset */
+  /* store copies of descriptors so they can be compared after the reset.
+     the mutex orders this read against the hotplug thread refreshing
+     dev_descriptor (under the same mutex) while an earlier re-enumeration
+     completes. */
+  usbi_mutex_lock(&darwin_cached_devices_mutex);
   memcpy (&descriptor, &dpriv->dev_descriptor, sizeof (descriptor));
+  usbi_mutex_unlock(&darwin_cached_devices_mutex);
   cached_configurations = (IOUSBConfigurationDescriptor *)alloca (sizeof (*cached_configurations) * descriptor.bNumConfigurations);
 
   for (i = 0 ; i < descriptor.bNumConfigurations ; ++i) {
@@ -2290,6 +2357,10 @@ static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, b
       return LIBUSB_ERROR_TIMEOUT;
     }
   }
+
+  /* the hotplug thread has processed the re-attach: adopt the replacement
+     device interface and service it discovered */
+  darwin_adopt_pending_device (dpriv);
 
   /* compare descriptors */
   usbi_dbg (ctx, "darwin/reenumerate_device: checking whether descriptors changed");
@@ -3000,6 +3071,11 @@ static int darwin_detach_kernel_driver_locked (struct libusb_device_handle *dev_
   if (get_interface_interface_version() < 700) {
     return LIBUSB_ERROR_NOT_SUPPORTED;
   }
+
+  /* adopt any replacement left over from a re-enumeration that timed out:
+     IOServiceAuthorize and darwin_reload_device below must operate on the
+     live service */
+  darwin_adopt_pending_device (dpriv);
 
   if (dpriv->capture_count == 0) {
     usbi_dbg (ctx, "attempting to detach kernel driver from device");

--- a/libusb/os/darwin_usb.h
+++ b/libusb/os/darwin_usb.h
@@ -116,6 +116,13 @@ struct darwin_cached_device {
   char                  sys_path[21];
   usb_device_t          device;
   io_service_t          service;
+  /* replacement interface and service discovered by the hotplug thread while
+     the device is re-enumerating. adopted (and the old ones released) by the
+     re-enumerating thread once the re-enumeration completes, so device and
+     service themselves are never written by the hotplug thread.
+     GUARDED_BY(darwin_cached_devices_mutex) */
+  usb_device_t          pending_device;
+  io_service_t          pending_service;
   usbi_mutex_t          lock;          /* protects open_count and capture_count */
   int                   open_count;    /* GUARDED_BY(lock) */
   int                   capture_count; /* GUARDED_BY(lock) */

